core/src/main/java/io/grpc/Metadata.java

```diff
@@ -396,14 +413,56 @@ private String toStringInternal() {
       return new AsciiKey<T>(name, marshaller);
     }
 
+    private final String originalName;
+
     private final String name;
-    private final byte[] asciiName;
+    private final byte[] nameBytes;
+
+    private static BitSet generateValidTChars() {
+      BitSet valid  = new BitSet(0x7f);
+      valid.set('-');
+      valid.set('_');
+      for (char c = '0'; c <= '9'; c++) {
+        valid.set(c);
+      }
+      // Only validates after normalization, so we exclude uppercase.
+      for (char c = 'a'; c <= 'z'; c++) {
+        valid.set(c);
+      }
+      return valid;
+    }
+
+    private static String validateName(String n) {
+      checkNotNull(n);
+      checkArgument(n.length() != 0, "token must have at least 1 tchar");
+      for (int i = 0; i < n.length(); i++) {
+        char tChar = n.charAt(i);
+        // TODO(notcarl): remove this hack once pseudo headers are properly handled
+        if (tChar == ':' && i == 0) {
+          continue;
+        }
+
+        checkArgument(VALID_T_CHARS.get(tChar), "Invalid character '%s' in key name", tChar);
+      }
+      return n;
+    }
 
     private Key(String name) {
-      this.name = Preconditions.checkNotNull(name, "name").toLowerCase(Locale.ROOT).intern();
```

Comment:
I'm fine with the intern(). I wouldn't have put it there if I wrote the code, but it didn't seem bad enough to remove. I'd prefer to not change such behavior in a PR that is just supposed to be adding validation.

That said, the benefit of intern() would be faster equals() checking if you have two Key instances with the same name, since the references will be the same.